test/distributed/org/apache/cassandra/distributed/test/sai/PartialUpdateHandlingTest.java

```diff
@@ -0,0 +1,541 @@
+
+            for (Object column : specification.nonKeyColumns())
+                select.append(", ").append(column);
+
+            select.append(" FROM ").append(KEYSPACE).append('.').append(specification.tableName()).append(" WHERE ");
+
+            ArrayList<String> restricted = Lists.newArrayList(specification.columns);
+
+            if (specification.restrictPartitionKey)
+            {
+                restricted.add("pk");
+                restricted.add("pk2");
+            }
+
+            List<String> clauses = new ArrayList<>();
+            boolean needsAllowFiltering = false;
+
+            if (specification.validationMode == EQ)
+            {
+                Map<String, Integer> primaryRow = modelRows.get(0);
+                assertEquals(specification.partitionKey, primaryRow.get("pk").intValue());
+                
+                for (String column : restricted)
+                {
+                    clauses.add(column + " = " + primaryRow.get(column));
+                    needsAllowFiltering |= isNotIndexed(column);
+                }
+            }
+            else if (specification.validationMode == RANGE)
+            {
+                // Attempt to match the first half of the model's rows...
+                for (String column : restricted)
+                {
+                    int min = modelRows.get(0).get(column);
+                    clauses.add(column + " >= " + min);
+                    int max = modelRows.get(PARTITIONS_PER_TEST / 2).get(column);
+                    clauses.add(column + " < " + max);
+
+                    needsAllowFiltering |= isNotIndexed(column);
+                }
+            }
+            else
+                throw new IllegalStateException("Validation mode must be EQ or RANGE");
+
+            select.append(String.join(" AND ", clauses));
+
+            if (needsAllowFiltering)
+                select.append(" ALLOW FILTERING");
+
+            InetAddressAndPort node2AddressAndPort = InetAddressAndPort.getByAddress(CLUSTER.get(2).broadcastAddress());
+            Map<String, Index.Status> node2IndexStatusFrom1;
+            Map<String, Index.Status> node2IndexStatusFrom3;
+
+            // Make sure node 2, which is the only complete replica, can't participate in a QUORUM...
+            if (specification.readCL == QUORUM)
+            {
+                node2IndexStatusFrom1 = CLUSTER.get(1).callsOnInstance(() -> IndexStatusManager.instance.peerIndexStatus.remove(node2AddressAndPort)).call();
+                node2IndexStatusFrom3 = CLUSTER.get(3).callsOnInstance(() -> IndexStatusManager.instance.peerIndexStatus.remove(node2AddressAndPort)).call();
+            }
+            else
+            {
+                node2IndexStatusFrom1 = null;
+                node2IndexStatusFrom3 = null;
+            }
+
+            Object[][] result = CLUSTER.coordinator(1).execute(select.toString(), specification.readCL);
```

Comment:
There might be some value in also using `executeWithPaging` here. I remember paging being tricky for the RFP/SRP combo during 8272.